rootchain/samuel/samuel.go

```diff
@@ -0,0 +1,580 @@
+package samuel
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+
+	"github.com/0xPolygon/polygon-edge/crypto"
+	"github.com/0xPolygon/polygon-edge/e2e/framework"
+	"github.com/0xPolygon/polygon-edge/rootchain"
+	"github.com/0xPolygon/polygon-edge/rootchain/payload"
+	"github.com/0xPolygon/polygon-edge/rootchain/proto"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/hashicorp/go-hclog"
+	"github.com/umbracle/ethgo/abi"
+	googleProto "google.golang.org/protobuf/proto"
+)
+
+var (
+	errUnknownPayloadType = errors.New("unknown payload type")
+)
+
+// eventTracker defines the event tracker interface for SAMUEL
+type eventTracker interface {
+	// Start starts the event tracker from the specified block number
+	Start(uint64) error
+
+	// Stop stops the event tracker
+	Stop() error
+
+	// Subscribe creates a rootchain event subscription
+	Subscribe() <-chan rootchain.Event
+}
+
+// samp defines the SAMP interface for SAMUEL
+type samp interface {
+	// AddMessage pushes a Signed Arbitrary Message into the SAMP
+	AddMessage(rootchain.SAM) error
+
+	// Prune prunes out all SAMs based on the specified event index
+	Prune(uint64)
+
+	// Peek returns a ready set of SAM messages, without removal
+	Peek() rootchain.VerifiedSAM
+
+	// Pop returns a ready set of SAM messages, with removal
+	Pop() rootchain.VerifiedSAM
+}
+
+// signer defines the signer interface used for
+// generating signatures
+type signer interface {
+	// Sign signs the specified data,
+	// and returns the signature and the block number at which
+	// the signature was generated
+	Sign([]byte) ([]byte, uint64, error)
+
+	// VerifySignature verifies the signature for the passed in
+	// raw data, and at the specified block number
+	VerifySignature([]byte, []byte, uint64) error
+
+	// Quorum returns the number of quorum validators
+	// for the given block number
+	Quorum(uint64) uint64
+}
+
+// transport defines the transport interface used for
+// publishing and subscribing to gossip events
+type transport interface {
+	// Publish gossips the specified SAM message
+	Publish(*proto.SAM) error
+
+	// Subscribe subscribes for incoming SAM messages
+	Subscribe(func(*proto.SAM)) error
+}
+
+// storage defines the required storage interface for SAMUEL
+// and its modules
+type storage interface {
+	// ReadLastProcessedEvent reads the last processed event data
+	ReadLastProcessedEvent(string) (string, bool)
+
+	// WriteLastProcessedEvent writes the last processed event data
+	WriteLastProcessedEvent(data string, contractAddr string) error
+}
+
+// eventData holds information on event data mapping
+type eventData struct {
+	payloadType  rootchain.PayloadType
+	eventABI     *abi.Event
+	methodABI    *abi.Method
+	localAddress types.Address
+}
+
+// SAMUEL is the module that coordinates activities with the SAMP and Event Tracker
+type SAMUEL struct {
```

Comment:
Please explain what SAMUEL means so that everyone that has not read the Bible can understand it.

Reply:
We will have detailed explanations on the SAMUEL module after we complete the effort.
Right now, it doesn't make sense to go into detail on what it does in the comments

Reply:
Keeping this comment unresolved to remember what we talk here

Reply:
Resolving because documentation on SAMUEL will be added to the RFC later on